File: src/_plugins/cleanup_text.rb
# frozen_string_literal: true

# Provides a filter that "cleans up" text.
#
# In particular this inserts HTML entities to prevent text wrapping
# in unfortunate places.  e.g. 'RFC 1234' becomes 'RFC&nbsp;1234'
# Doing this in a global filter means:
#
#   1.  These rules are applied consistently
#   2.  I don't have to litter my Markdown source with non-breaking
#       HTML entities
#
# See: https://alexwlchan.net/2020/adding-non-breaking-spaces-with-jekyll/

def add_non_breaking_spaces(input)
  text = input

  # "e.g." introduces an example; it's annoying when the example text has
  # been bumped to another line, so don't let it.
  text.gsub('e.g. ', 'e.g.&nbsp;')

  # Add a non-breaking space after words which are followed by
  # a number, e.g. 'Apollo 11' or 'RFC 456'
  prefix_words = %w[
    Apollo
    Artemis
    chapter
    HTTP
    ImageMagick
    ISO/IEC
    Issue
    issue
    iPres
    No.
    Part
    part
    Python
    RFC
    Safari
    Season
    season
    SQLite
  ].join('|')

  text = text.gsub(/(#{prefix_words}) (\d+)/, '\1&nbsp;\2')

  # Add a non-breaking space after words which are preceded by
  # a number, e.g. '1 second' or '5 bytes'
  countable_words = %w[
    bookmark
    byte
    character
    count
    GB
    hour
    inch
    kilometre
    line
    MiB
    million
    minute
    second
    tags
    unit
    vote
    year
  ].join('|')

  text = text.gsub(/(\d+) (#{countable_words})/, '\1&nbsp;\2')

  # Other phrases which needed non-breaking spaces or non-breaking
  # dashes.
  phrases = [
    '<em>k</em>-means',
    '26k items',
    'Algorithm L',
    'Algorithm R',
    'Amazon S3',
    'Apple TV+',
    'CC0 1.0',
    'CC BY 2.0',
    'CC BY 3.0',
    'CC BY 4.0',
    'CC BY-NC 2.0',
    'CC BY-NC 4.0',
    'CC BY-ND 2.0',
    'CC BY-NC-ND',
    'CC BY-SA 2.0',
    'CC BY-SA 3.0',
    'CC BY-SA 4.0',
    'CC BY-NC-SA 4.0',
    'CC BY',
    'DjangoCon US',
    'Dr. Drang',
    'ECMA-404',
    'Git LFS',
    'HTTP 200 OK',
    'iMac G3',
    'iPhone X',
    'IP address',
    'JPEG 2000',
    'Latin-1',
    'Mac OS 9',
    'Mac OS X',
    'Monki Gras',
    'MS Paint',
    'Objective-C',
    'P-215',
    'PDF 1.6',
    'PDF 1.7',
    'PyCon ',
    'Route 53',
    'Silo 49',
    'System 1',
    'Windows-1252',
    'z-axis'
  ]

  phrases.each do |p|
    if text.include? p
      replacement = p.gsub(' ', '&nbsp;').gsub('-', '&#8209;')
      text = text.gsub(p, replacement)
    end
  end

  # If there's an <a> that starts with the word "a", make sure we
  # add a non-breaking space to it.
  #
  #     <a href="https://example.com">an example</a>
  #  ~> <a href="https://example.com">an&nbsp;example</a>
  #
  text = text.gsub(/<a ([^>]+)>a /, '<a \1>a&nbsp;')
  text.gsub(/<a ([^>]+)>an /, '<a \1>an&nbsp;')
end

# The syntax highlighter wraps highlighted code blocks in:
#
#     <div class="language-{language} highlighter-rouge">
#       <div class="highlight">
#         <pre>
#           …
#
# We can strip out some of these CSS classes and HTML tags.
#
def cleanup_syntax_highlighter_classes(html)
  # I never use the `highlighter-rouge` class, and I only have styles
  # for a couple of the `language-*` classes.
  %w[caddy console go irb].each do |lang|
    html = html.gsub(" class=\"language-#{lang} highlighter-rouge\"", " class=\"language-#{lang}\"")
  end

  html = html.gsub(/ class="language-[a-z]+ highlighter-rouge"/, '')

  # I never use the `highlight` class.
  #
  # If there's a `language-tag` on the outer `<div>`, move it to the `<pre>`.
  html = html.gsub(
    /<div class="language-(?<language>[a-z]+)"><div class="highlight"><pre class="highlight">/,
    '<pre class="language-\k<language>">'
  )
  html = html.gsub(
    '<div><div class="highlight"><pre class="highlight">',
    '<pre>'
  )
  html = html.gsub(%r{</pre></div>\s*</div>}, '</pre>')
  html = html.gsub(
    /<code class="language-(?<language>[a-z]+)" data-lang="[a-z]+">/,
    '<code>'
  )

  # Remove any whitespace before/after `<pre>` blocks
  html = html.gsub(/\s+<pre>/, '<pre>')
  html.gsub(%r{</pre>\s+}, '</pre>')
end

module Jekyll
  module CleanupsFilter
    def cleanup_text(input)
      cache = Jekyll::Cache.new('CleanupText')

      cache.getset(input) do
        text = add_non_breaking_spaces(input)

        if text.include? '<pre'
          text = cleanup_syntax_highlighter_classes(text)
        end

        # Display "LaTeX" in a nice way, if you have CSS enabled
        if text.include? 'TeX'
          text = text.gsub(
            ' LaTeX',
            ' <span class="visually-hidden">LaTeK</span>' \
            '<span class="latex" aria-hidden="true">L<sup>a</sup>T<sub>e</sub>X</span>'
          )

          text = text.gsub(
            ' TeX',
            ' <span class="visually-hidden">TeK</span>' \
            '<span class="latex" aria-hidden="true">T<sub>e</sub>X</span>'
          )
        end

        # Make sure that footnote markers are rendered as a text
        # arrow on iOS devices, not emoji.  For more info:
        # http://daringfireball.net/linked/2015/04/22/unicode-emoji
        text = text
               .gsub('&#8617;', '&#8617;&#xFE0E;')
               .gsub('↩', '&#8617;&#xFE0E;')

        text.strip
      end
    end
  end
end

Liquid::Template.register_filter(Jekyll::CleanupsFilter) if defined? Liquid

File: mosaic/text/cleanup_text.py
"""
Utilities for dealing with text.
"""

import os
import re


from mosaic import cache


__all__ = ["cleanup_text"]


# Use the mtime of this file as part of the cache key, to ensure that
# when the definitions in this file change, the calues are recalculated
CLEANUP_TEXT_CACHE_ID = str(os.stat(__file__).st_mtime)


def cleanup_text(text: str) -> str:
    """
    Apply all my cleanup rules to text.
    """
    cache_ns = "cleanup_text"
    cache_key = f"{CLEANUP_TEXT_CACHE_ID}:{cache.md5(text)}"

    if value := cache.get(cache_ns, cache_key):
        return value

    text = add_non_breaking_characters(text)
    text = add_latex_css_classes(text)
    text = force_text_footnote_markers(text)
    text = text.strip()

    cache.set(cache_ns, cache_key, text)
    return text


# Words which often appear before a number
PREFIX_WORDS = [
    "Apollo",
    "Artemis",
    "chapter",
    "HTTP",
    "ImageMagick",
    "ISO/IEC",
    "Issue",
    "issue",
    "iPres",
    "Mr.",
    "Mrs.",
    "No.",
    "Part",
    "part",
    "Python",
    "RFC",
    "Safari",
    "Season",
    "season",
    "SQLite",
    "Thunderbolt",
    "Xcode",
]

# Construct a regex like `(issue|part|season) (\d+)`.
PREFIX_RE = re.compile("(" + "|".join(PREFIX_WORDS) + r") (\d+)")

# Words which can be counted, which often follow a number
COUNTABLE_WORDS = [
    "AD",
    "BC",
    "bookmark",
    "byte",
    "character",
    "count",
    "GB",
    "GiB",
    "hour",
    "inch",
    "kilometre",
    "line",
    "MiB",
    "million",
    "millisecond",
    "minute",
    "second",
    "tags",
    "unit",
    "vote",
    "year",
]

# Construct a regex like `(\d+) (byte|hour|inch)`
COUNTABLE_RE = re.compile(r"(\d+) (" + "|".join(COUNTABLE_WORDS) + ")")

NON_BREAKING_PHRASES = [
    "<em>k</em>-means",
    "200 OK",
    "26k items",
    "Algorithm L",
    "Algorithm R",
    "Amazon S3",
    "Apple TV+",
    "CC0 1.0",
    "CC BY 2.0",
    "CC BY 3.0",
    "CC BY 4.0",
    "CC BY-NC 2.0",
    "CC BY-NC 4.0",
    "CC BY-ND 2.0",
    "CC BY-NC-ND",
    "CC BY-SA 2.0",
    "CC BY-SA 3.0",
    "CC BY-SA 4.0",
    "CC BY-NC-SA 4.0",
    "CC BY",
    "C.S. Lewis",
    "DjangoCon US",
    "Dr. Drang",
    "ECMA-404",
    "e.g. ",
    "Face ID",
    "Flickr API",
    "Git LFS",
    "HTTP 200 OK",
    "iMac G3",
    "iPhone X",
    "IP address",
    "JPEG 2000",
    "Latin-1",
    "Mac OS 9",
    "Mac OS X",
    "Monki Gras",
    "MS Paint",
    "New York",
    "Objective-C",
    "P-215",
    "PDF 1.6",
    "PDF 1.7",
    "PyCon ",
    "Route 53",
    "Silo 49",
    "System 1",
    "Touch ID",
    "UTF-8",
    "UTF-16",
    "UTF-32",
    "VS Code",
    "Windows-1252",
    "z-axis",
]

PROPER_NAME_RE = re.compile(r"(?P<initials>[A-Z]\.[A-Z]\.) (?P<surname>[A-Z])")

# CODE_FLAG_RE matches <code> snippets that are short.
#
# The limit of 15 characters is arbitrary. In longer code snippets,
# wrapping is preferable to avoid leaving excessive whitespace on
# the previous line.
CODE_FLAG_RE = re.compile(r"<code>(?P<contents>[^<]{1,15})</code>")


def add_nowrap(match: re.Match[str]) -> str:
    """
    Add the `nowrap` class to a `<code>` element if it contains line
    breaking characters.
    """
    contents = match.group("contents")
    if "-" in contents or " " in contents:
        return f'<code class="nowrap">{contents}</code>'
    return match.group(0)


def add_non_breaking_characters(text: str) -> str:
    """
    Add non-breaking spaces and characters to my text.

    See https://alexwlchan.net/2020/adding-non-breaking-spaces-with-jekyll/
    """
    # Add a non-breaking space after words which are followed by a number,
    # like "part 5" or "issue 123".
    text = PREFIX_RE.sub(r"\1&nbsp;\2", text)

    # Add a non-breaking space before words which are preceded by a number,
    # like "1 byte" or "4 inches".
    text = COUNTABLE_RE.sub(r"\1&nbsp;\2", text)

    # Add a non-breaking space after words if they're the first word
    # in a sentence.
    short_words = ["A", "An", "I"]

    for w in short_words:
        text = text.replace(f". {w} ", f". {w}&nbsp;")
        text = text.replace(f".\n{w} ", f".\n{w}&nbsp;")

    # Add a non-breaking space in phrases that look like names.
    text = PROPER_NAME_RE.sub(r"\g<initials>&nbsp;\g<surname>", text)

    # Handle other phrases which need non-breaking spaces or dashes.
    for phrase in NON_BREAKING_PHRASES:
        if phrase not in text:
            continue

        replacement = phrase.replace(" ", "&nbsp;").replace("-", "&#8209;")
        text = text.replace(phrase, replacement)

    # Add a `nowrap` class to short <code> snippets which contain
    # potential line-break characters, so they don't wrap.
    text = CODE_FLAG_RE.sub(add_nowrap, text)

    return text


def add_latex_css_classes(html: str) -> str:
    """
    Make the words TeX and LaTeX display nicely in my HTML.
    """
    latex_css = '<link href="css/components/latex.css" rel="stylesheet"/>'
    latex_hidden = '<span class="visually-hidden">LaTeK</span>'
    latex_visual = (
        '<span class="latex" aria-hidden="true">L<sup>a</sup>T<sub>e</sub>X</span>'
    )

    if html == "LaTeX":
        return f"{latex_css}{latex_hidden}{latex_visual}"

    html = html.replace(
        " LaTeX",
        f"{latex_css}{latex_hidden} {latex_visual}",
    )
    html = html.replace(">LaTeX<", f">{latex_css}{latex_hidden}{latex_visual}<")
    html = html.replace(">LaTeX ", f">{latex_css}{latex_hidden}{latex_visual} ")

    html = html.replace(
        " TeX",
        latex_css + '<span class="visually-hidden">TeK</span>'
        ' <span class="latex" aria-hidden="true">T<sub>e</sub>X</span>',
    )

    return html


def force_text_footnote_markers(html: str) -> str:
    """
    Force footnote markers to render as text on iOS devices, not emoji.
    """
    # See https://mts.io/2015/04/21/unicode-symbol-render-text-emoji/
    return html.replace("&#8617;", "&#8617;&#xFE0E;").replace("↩", "&#8617;&#xFE0E;")
